### PR TITLE
New operator 'expand'

### PR DESCRIPTION
**Pull request by zrphercule**
New operator expand, behave like numpy.broadcast.
Documents & more testcases will be submitted once testing finished.

### CODE REVIEW

**Comment by zrphercule** [comment]
@ezyang yes, this is materialized, unlike numpy or pytorch.expand

File: caffe2/operators/expand_op.h [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+      : Operator<Context>(operator_def, ws) {}
+
+  bool RunOnDevice() override {
+    return DispatchHelper<InputTypes>::call(this, Input(0));
+  }
+ template <typename T>
+  bool DoRunWithType() {
+    const auto& X = Input(0);
+    const auto& Y_shape_tensor = Input(1);
+	std::vector<int> shape_dims(Y_shape_tensor.size());
```
Comment (bddppq) [inline caffe2/operators/expand_op.h]: `int64_t`

File: caffe2/operators/expand_op.h [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+	context_.template Copy<int, Context, CPUContext>(Y_shape_tensor.size(), Y_shape_tensor.template data<int>(), shape_dims.data());
+    const int* Y_shape = Y_shape_tensor.template data<int>();
```
Comment (bddppq) [inline caffe2/operators/expand_op.h]: nit: move this to one line above so the `Copy` call can use this variable

File: caffe2/operators/expand_op.h [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    auto* Y = Output(0);
+
+	const int ndim = shape_dims.size();
```
Comment (bddppq) [inline caffe2/operators/expand_op.h]: ident

File: caffe2/operators/expand_op.h [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+    const std::vector<int> X_dims(X.dims().cbegin(), X.dims().cend());
+    std::vector<int> Y_dims;
+    Y_dims.reserve(std::max(ndim, X.ndim()));
+    // ndim, X.ndim() might equal to 0
+    std::cout << "loginfoo " << ndim << " " << X.ndim() << std::endl;
+    for (int i = ndim - 1, j = X.ndim() - 1; i >= 0 || j >= 0; i--, j--) {
```
Comment (bddppq) [inline caffe2/operators/expand_op.h]: looks to me this is quite similar to https://github.com/pytorch/pytorch/blob/d1bdb3b10ae789729e8aef48d85b4835f9bdfd51/caffe2/utils/math_cpu.cc#L876, could you refactor to use that util function? it's better to keep the broadcast implementation the same across all the places.

File: caffe2/operators/expand_op.h [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    return DispatchHelper<InputTypes>::call(this, Input(0));
+  }
+
+  template <typename T>
+  bool DoRunWithType() {
+    const auto& dY = Input(0);
+    const auto& X = Input(1);
+    const auto& Y = Input(2);
+    auto* dX = Output(0);
+    const int ndim = Y.ndim();
```
Comment (BIT-silence) [inline caffe2/operators/expand_op.h]: It seems there is no need for Y as one of the inputs since we only use Y for the dims which we can get from dY.

File: caffe2/operators/expand_op.h [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+        X_dims,
+        Y_dims.size(),
+        Y_dims,
+        X.template data<T>(),
+        Y->template mutable_data<T>(),
+        &context_);
+  }
+
+ private:
+  Expander expander_{};
```
Comment (BIT-silence) [inline caffe2/operators/expand_op.h]: If we don't have any plan for some other expand ops. I this there is no need for the Expander part. It is kind of over design. I'd suggest make the logic of Forward and Backward   in RunOnDevice in ExpandOp and ExpandGradientOp directly.

File: caffe2/operators/expand_op_gpu.cc [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+#include <vector>
+
+#include "caffe2/core/common_gpu.h"
+#include "caffe2/core/context_gpu.h"
+
+namespace caffe2 {
+
+REGISTER_CUDA_OPERATOR(
+    Expand,
+
```
Comment (BIT-silence) [inline caffe2/operators/expand_op_gpu.cc]: Empty line?

File: caffe2/operators/expand_op.cc [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+namespace {
+
+class GetExpandGradient final : public GradientMakerBase {
+  using GradientMakerBase::GradientMakerBase;
+  std::vector<OperatorDef> GetGradientDefs() override {
+    return SingleGradientDef(
+        def_.type() + "Gradient",
+        "",
+        std::vector<string>{GO(0), I(0), O(0)},
```
Comment (BIT-silence) [inline caffe2/operators/expand_op.cc]: If there is no need for Y as part of Gradient Input, remove O(0) here.

File: caffe2/operators/expand_op.h [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+ private:
+  const Expander expander_{};
+};
+
+template <class Context>
+struct Expander {
+  template <typename T>
+  bool Forward(
+      const int& X_ndims,
+      const std::vector<int>& X_dims,
```
Comment (BIT-silence) [inline caffe2/operators/expand_op.h]: If use vector here, the vector::size will be equal to X_ndim, right? So consider simplify the function definition.

File: caffe2/operators/expand_op.h [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+    const auto& Y_shape_tensor = Input(1);
+	std::vector<int> shape_dims(Y_shape_tensor.size());
+	context_.template Copy<int, Context, CPUContext>(Y_shape_tensor.size(), Y_shape_tensor.template data<int>(), shape_dims.data());
+    const int* Y_shape = Y_shape_tensor.template data<int>();
+    auto* Y = Output(0);
+
+	const int ndim = shape_dims.size();
+    const std::vector<int> X_dims(X.dims().cbegin(), X.dims().cend());
+    std::vector<int> Y_dims;
+    Y_dims.reserve(std::max(ndim, X.ndim()));
```
Comment (BIT-silence) [inline caffe2/operators/expand_op.h]: I'd suggest just allocate Y_dims here and use operator[] instead of push back in the following lines. This is just some personal advice.

File: caffe2/operators/expand_op.h (hunk 4 above)
Comment (BIT-silence) [reply caffe2/operators/expand_op.h]: Remove std::cout here.

File: caffe2/operators/expand_op.h [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+  const Expander expander_{};
+};
+
+template <class Context>
+struct Expander {
+  template <typename T>
+  bool Forward(
+      const int& X_ndims,
+      const std::vector<int>& X_dims,
+      const int& Y_ndims,
```
Comment (BIT-silence) [inline caffe2/operators/expand_op.h]: const int Y_ndims is fine. For built-in types without template, there is no need to pass by reference.

File: caffe2/operators/expand_op_gpu.cc [hunk 12]
```diff
@@ -0,0 +1,25 @@
+#include "caffe2/operators/expand_op.h"
+
+#include <algorithm>
+#include <functional>
+#include <vector>
+
+#include "caffe2/core/common_gpu.h"
```
Comment (BIT-silence) [inline caffe2/operators/expand_op_gpu.cc]: No need to include this?

File: caffe2/operators/expand_op_gpu.cc [hunk 13]
```diff
@@ -0,0 +1,25 @@
+#include "caffe2/operators/expand_op.h"
+
+#include <algorithm>
```
Comment (BIT-silence) [inline caffe2/operators/expand_op_gpu.cc]: No need to include these std headers? It seems we don't use them in this file.

File: caffe2/python/operator_test/expand_op_test.py [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+            j -= 1
+        shape = np.array(shape_list, dtype=np.int32)
+
+        op = core.CreateOperator(
+            op_name,
+            ["X", "shape"],
+            ["Y"],
+        )
+        def ref(X, shape):
+            return (np.array(X) * np.ones(shape),)
```
Comment (BIT-silence) [inline caffe2/python/operator_test/expand_op_test.py]: I'm not sure if this is the best practice, consider using np.tile?

File: caffe2/operators/expand_op.h (hunk 2 above)
Comment (zrphercule) [reply caffe2/operators/expand_op.h]: Infact I can just delete this var since it is no use anymore...

File: caffe2/operators/expand_op.h (hunk 11 above)
Comment (zrphercule) [reply caffe2/operators/expand_op.h]: yes, I wont use X_ndims or Y_ndims anymore; as you suggested, I will use size of vector.

File: caffe2/operators/expand_op_gpu.cc (hunk 12 above)
Comment (zrphercule) [reply caffe2/operators/expand_op_gpu.cc]: I believe so

File: caffe2/python/operator_test/expand_op_test.py (hunk 14 above)
Comment (zrphercule) [reply caffe2/python/operator_test/expand_op_test.py]: I think np.broadcast_to is equals to this function?

File: caffe2/operators/expand_op.cc [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+
+OPERATOR_SCHEMA(ExpandGradient).NumInputs(2).NumOutputs(1);
+
+namespace {
+
+class GetExpandGradient final : public GradientMakerBase {
+  using GradientMakerBase::GradientMakerBase;
+  std::vector<OperatorDef> GetGradientDefs() override {
+    return SingleGradientDef(
+        def_.type() + "Gradient",
```
Comment (BIT-silence) [inline caffe2/operators/expand_op.cc]: Here I'd suggest to use "ExpandGradient" directly.

File: caffe2/python/operator_test/expand_op_test.py (hunk 14 above)
Comment (BIT-silence) [reply caffe2/python/operator_test/expand_op_test.py]: This will make things clear especially in the test. Use np.broadcast_to will let us know that what we are testing is to make sure the op has the same functionality of np.broadcast_to.

File: caffe2/operators/expand_op.h [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+    return expander_.template Forward<T>(
+        X_dims,
+        Y_dims,
+        X.template data<T>(),
+        Y->template mutable_data<T>(),
+        &context_);
+  }
+
+ private:
+  Expander<Context> expander_{};
```
Comment (BIT-silence) [inline caffe2/operators/expand_op.h]: I mean remove the Expander struct at all. There is no need to be such a functor template for ExpandOp.

Comment (zrphercule) [reply caffe2/operators/expand_op.h]: OK Boss!

File: caffe2/python/operator_test/expand_op_test.py (hunk 14 above)
Comment (zrphercule) [reply caffe2/python/operator_test/expand_op_test.py]: @BIT-silence in fact we cannot use np.broadcast_to here, since our function is a little bit different from np.broadcast_to.

for example:

we have an array X, X.shape = [3,1]
let Y = caffe2.Expand(X, (3,1,3))
then Y.shape = [3,3,3], and that's what we want.

however, if we
let Y = np.broadcast_to(X, (3,1,3)),
then np will raise a value error notify you that the second dimension of new_shape is smaller than X.shape, and it is illegal.

Comment (BIT-silence) [reply caffe2/python/operator_test/expand_op_test.py]: I see, thanks for the clarification.

**Review by xiaomengy — APPROVED** [review]

File: caffe2/operators/expand_op.h [hunk 17]
```diff
@@ -0,0 +1,107 @@
+#ifndef CAFFE2_OPERATORS_REDUCE_OPS_H_
+#define CAFFE2_OPERATORS_REDUCE_OPS_H_
+
+#include <algorithm>
+#include <functional>
```
Comment (bddppq) [inline caffe2/operators/expand_op.h]: this doesn't seem needed anymore?

File: caffe2/operators/expand_op.h [hunk 18]
```diff
@@ -0,0 +1,107 @@
...
+        Y_shape_tensor.template data<int64_t>(),
+        shape_dims.data());
+    auto* Y = Output(0);
+
+	const int ndim = shape_dims.size();
+    const std::vector<int> X_dims(X.dims().cbegin(), X.dims().cend());
+    std::vector<int> Y_dims;
+    Y_dims.reserve(std::max(ndim, X.ndim()));
+    // ndim, X.ndim() might equal to 0
+    for (int i = ndim - 1, j = X.ndim() - 1; i >= 0 || j >= 0; i--, j--) {
```
Comment (bddppq) [inline caffe2/operators/expand_op.h]: This block should be as simple as copying the Input(1) to Y_dims, the validation work should be done in math::Broadcast already.

Comment (zrphercule) [reply caffe2/operators/expand_op.h]: Maybe not? Or we can have another function to calculate shape. I think it is not a very good idea to let math::Broadcast handle too many things, but better just let it accept some standard data?

File: caffe2/python/operator_test/expand_op_test.py [hunk 19]
```diff
@@ -0,0 +1,49 @@
...
+import itertools as it
+
+class TestExpandOp(hu.HypothesisTestCase):
+    def run_expand_op_test(
+        self, op_name, X, gc, dc):
+        shape_length = np.random.randint(5)
+        shape_list = []
+        j = shape_length-1
+        i = X.ndim-1
+        while i >= 0 or j >= 0:
```
[the remark this answers is not in the document]
Comment (zrphercule) [reply caffe2/python/operator_test/expand_op_test.py]: I think you are right. I will enumerate all possible shape_length from 1 to 5, instead of give it a random length.

[the remark this answers is not in the document]
Comment (zrphercule) [reply caffe2/python/operator_test/expand_op_test.py]: I will keep both these too versions, one random shape, one determined shape :)

**Comment by zrphercule** [comment]
@bddppq could you please help me to retest it? thanks!

File: caffe2/operators/expand_op.h [hunk 20]
```diff
@@ -0,0 +1,105 @@
...
```
Comment (bddppq) [inline caffe2/operators/expand_op.h]: --i, --j

File: caffe2/python/operator_test/expand_op_test.py [hunk 21]
```diff
@@ -0,0 +1,59 @@
...
+from __future__ import print_function
+from __future__ import unicode_literals
+
+from caffe2.python import core, workspace
+from hypothesis import given
+
+import caffe2.python.hypothesis_test_util as hu
+import hypothesis.strategies as st
+import numpy as np
+import itertools as it
```
Comment (bddppq) [inline caffe2/python/operator_test/expand_op_test.py]: unused

File: caffe2/python/operator_test/expand_op_test.py [hunk 22]
```diff
@@ -0,0 +1,59 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+from caffe2.python import core, workspace
+from hypothesis import given
+
+import caffe2.python.hypothesis_test_util as hu
+import hypothesis.strategies as st
```
Comment (bddppq) [inline caffe2/python/operator_test/expand_op_test.py]: unused

File: caffe2/python/operator_test/expand_op_test.py [hunk 23]
```diff
@@ -0,0 +1,59 @@
...
+import numpy as np
+import itertools as it
+
+
+class TestExpandOp(hu.HypothesisTestCase):
+    def run_expand_op_test(
+        self, op_name, X, random_flag, gc, dc):
+        if (random_flag):
```
Comment (bddppq) [inline caffe2/python/operator_test/expand_op_test.py]: remove parentheses

**Review by bddppq — APPROVED** [review]
Pre-approving

As discussed offline, please add couple test cases with explicit (non-random) shapes.
